core/src/main/scala/org/http4s/multipart/MultipartParser.scala

```diff
@@ -388,134 +387,186 @@ object MultipartParser {
-            }
-            .pull
-            .echo *> r.pull.uncons.flatMap {
-            case Some(_) =>
-              tailrecParse(r, headers)
-            case None =>
-              Pull.done
-          }
-      }
-
-    tailrecParse(strim, Headers.empty).stream
-      .fold(Headers.empty)(_ ++ _)
-  }
-
-  private def streamStageIgnoreRest[F[_]: Sync](
+  /**
+    *
+    * @param boundary
+    * @param s
+    * @param limit
+    * @tparam F
+    * @return
+    */
+  private def pullParts[F[_]: Sync](
       boundary: Boundary,
       s: Stream[F, Byte],
       limit: Int
-  ): Stream[F, Part[F]] = {
-    val endlineBytes = EndlineBytesN(boundary)
-    val values = ExpectedBytesN(boundary)
-    splitOrFail[F](endlineBytes, s).flatMap {
-      case (l, _) =>
-        streamStageParsePart[F](boundary, values, l ++ Stream.chunk(Chunk.bytes(values)), limit).pull.echo
-    }.stream
-  }
+  ): Pull[F, Part[F], Unit] = {
+    val values = DoubleCRLFBytesN
+    val expectedBytes = ExpectedBytesN(boundary)
 
-  private def streamStageParsePart[F[_]: Sync](
-      boundary: Boundary,
-      values: Array[Byte],
-      s: Stream[F, Byte],
-      limit: Int
-  ): Stream[F, Part[F]] =
-    splitHalf[F](values, s).flatMap {
+    splitOrFinish[F](values, s, limit).flatMap {
       case (l, r) =>
-        r.pull.unconsChunk.flatMap {
-          case None =>
-            parseToPartStreamed[F](l, limit).pull.echo *>
-              Pull.done
-          case Some(_) =>
-            tailrecParts[F](boundary, values, l, r, limit)
+        //We can abuse reference equality here for efficiency
+        //Since `splitOrFinish` returns `empty` on a capped stream
+        //However, we must have at least one part, so `splitOrFinish` on this function
+        //Indicates an error
+        if (r == streamEmpty) {
+          Pull.raiseError(MalformedMessageBodyFailure("Cannot parse empty stream"))
+        } else {
+          tailrecParts[F](boundary, l, r, expectedBytes, limit)
         }
-    }.stream
+    }
+  }
 
   private def tailrecParts[F[_]: Sync](
       b: Boundary,
-      values: Array[Byte],
-      last: Stream[F, Byte],
-      next: Stream[F, Byte],
-      limit: Int
-  ): Pull[F, Part[F], Unit] =
-    parseToPartStreamed[F](last, limit).pull.echo *> splitHalf[F](values, next).flatMap {
-      case (l, r) =>
-        r.pull.uncons.flatMap {
-          case None =>
-            Pull.done
-          case Some(_) =>
-            tailrecParts[F](b, values, l, r, limit)
+      headerStream: Stream[F, Byte],
+      rest: Stream[F, Byte],
+      expectedBytes: Array[Byte],
+      limit: Int): Pull[F, Part[F], Unit] =
+    Pull
+      .eval(parseHeaders(headerStream))
+      .flatMap { hdrs =>
+        splitHalf(expectedBytes, rest).flatMap {
+          case (l, r) =>
+            //We hit a boundary, but the rest of the stream is empty
+            //and thus it's not a properly capped multipart body
+            if (r == streamEmpty) {
```

Comment:
Forgive both reference equality abuses: This gets caught downstream if it happens that the reference equality check doesn't hold, but it improves the common case of bodies constructed normally from byte streams a la blaze